mava/systems/jax/system.py

```diff
@@ -0,0 +1,137 @@
+        """System design specifying the list of components to use.
+
+        Returns:
+            system callback components
+        """
+
+    def update(self, component: Any) -> None:
+        """Update a component that has already been added to the system.
+
+        Args:
+            component : system callback component
+            name : component name
+        """
+        comp = component()
+        name = comp.name
+        if name in list(self._design.__dict__.keys()):
+            self._design.__dict__[name] = component
+            config_feed = {name: comp.config}
+            self.config.update(**config_feed)
+        else:
+            raise Exception(
+                "The given component is not part of the current system.\
+                Perhaps try adding it instead using .add()."
+            )
+
+    def add(self, component: Any) -> None:
+        """Add a new component to the system.
+
+        Args:
+            component : system callback component
+            name : component name
+        """
+        comp = component()
+        name = comp.name
+        if name in list(self._design.__dict__.keys()):
+            raise Exception(
+                "The given component is already part of the current system.\
+                Perhaps try updating it instead using .update()."
+            )
+        else:
+            self._design.__dict__[name] = component
+            config_feed = {name: comp.config}
+            self.config.add(**config_feed)
+
+    def configure(self, **kwargs: Any) -> None:
+        """Configure system hyperparameters."""
+        self.config.build()
+        self.config.set_parameters(**kwargs)
+
+    def launch(
+        self,
+        num_executors: int,
+        nodes_on_gpu: List[str],
+        multi_process: bool = True,
+        name: str = "system",
+    ) -> None:
+        """Run the system.
+
+        Args:
+            config : system configuration including
+            num_executors : number of executor processes to run in parallel
+            nodes_on_gpu : which processes to run on gpu
+            multi_process : whether to run locally or distributed, local runs are
+                for debugging
+            name : name of the system
+        """
+        # build config is not already built
+        if not self.config._built:
+            self.config.build()
+
+        # update distributor config
+        self.config.set_parameters(
+            num_executors=num_executors,
+            nodes_on_gpu=nodes_on_gpu,
+            multi_process=multi_process,
+            name=name,
+        )
+
+        # get system config to feed to component list to update hyperparamete settings
+        system_config = self.config.get()
+
+        # update default system component configs
+        for component in self._design.__dict__.values():
+            self.components.append(component(system_config))
+
+        # Build system
+        self._builder = Builder(components=self.components)
```

Comment:
Maybe we can make the `Builder` class configurable/param? This can be on the passed on the launch on init. That way classes can do the same launch, just using a different builder e.g. `MockBuilder` in the test :)

Reply:
So the builder will be highly configurable but not in the sense of being passed in, i.e. having different builders fed to the system/launch, but instead the configuration will come from overriding the hooks inside a general builder class through the builder component callbacks.